File: AM_Program/AM_Shared/ISpectraFilter.vb
'*********************************************************************************************************
' Written by Dave Clark for the US Department of Energy 
' Pacific Northwest National Laboratory, Richland, WA
' Copyright 2006, Battelle Memorial Institute
' Created 06/07/2006
'
'*********************************************************************************************************

Option Strict On

Public Interface ISpectraFilter

	'*********************************************************************************************************
	'Defines minimum required functionality for classes that will filter spectra files
	'*********************************************************************************************************

#Region "Enums"
	Enum ProcessResults
		'Return values for MakeProcess and Abort functions
		SFILT_SUCCESS = 0				 ' Operation succeeded
		SFILT_FAILURE = -1				 ' Operation failed
		SFILT_NO_FILES_CREATED = -2		 ' Spectra filter operation didn't fail, but no output files were created
		SFILT_ABORTED = -3				 ' Spectra filter operation aborted
		SFILT_NO_SPECTRA_ALTERED = -4	 ' Spectra filter did not alter any spectra
	End Enum

	Enum ProcessStatus
		'Return value for status property
		SFILT_STARTING				'Plugin initialization in progress
		SFILT_RUNNING				'Plugin is attempting to do its job
		SFILT_COMPLETE				'Plugin successfully completed its job
		SFILT_ERROR					'There was an error somewhere
		SFILT_ABORTING				'An ABORT command has been received; plugin shutdown in progress
	End Enum
#End Region

#Region "Structures"
	Structure InitializationParams
		Dim SourceFolderPath As String
		Dim OutputFolderPath As String
		'' Unused: Dim MiscParams As Specialized.StringDictionary
		Dim DebugLevel As Integer
		Dim Logger As PRISM.Logging.ILogger
		Dim MgrParams As IMgrParams
		Dim JobParams As IJobParams
		Dim StatusTools As IStatusFile
	End Structure
#End Region

#Region "Properties"
	ReadOnly Property Status() As ISpectraFilter.ProcessStatus 'Allows calling program to get current status

	ReadOnly Property Results() As ISpectraFilter.ProcessResults  'Allows calling program to determine if DTA creation succeeded

	ReadOnly Property ErrMsg() As String  'Error message describing any errors encountered

	ReadOnly Property SpectraFileCount() As Integer	'Count of spectra files that remain after filtering
#End Region

#Region "Methods"
	Sub Setup(ByVal InitParams As InitializationParams)	 'Initializes parameters. Must be called before executing Start()

	Function Start() As ISpectraFilter.ProcessStatus  'Starts the spectra filter operation

	Function Abort() As ISpectraFilter.ProcessStatus  'Aborts spectra filter operation
#End Region

End Interface

File: AM_Program/AM_Shared/ISpectraFilter.cs
//*********************************************************************************************************
// Written by Dave Clark for the US Department of Energy 
// Pacific Northwest National Laboratory, Richland, WA
// Copyright 2006, Battelle Memorial Institute
// Created 06/07/2006
//
//*********************************************************************************************************

namespace AnalysisManagerBase
{

    #region "Enums"

    /// <summary>
    /// Return values for MakeProcess and Abort functions
    /// </summary>
    public enum ProcessResults
    {
        /// <summary>
        /// Operation succeeded
        /// </summary>
        SFILT_SUCCESS = 0,

        /// <summary>
        /// Operation failed
        /// </summary>
        SFILT_FAILURE = -1,

        /// <summary>
        /// Spectra filter operation didn't fail, but no output files were created
        /// </summary>
        SFILT_NO_FILES_CREATED = -2,

        /// <summary>
        /// Spectra filter operation aborted
        /// </summary>
        SFILT_ABORTED = -3,

        /// <summary>
        /// Spectra filter did not alter any spectra
        /// </summary>
        SFILT_NO_SPECTRA_ALTERED = -4
        
    }

    /// <summary>
    /// Return value for status property
    /// </summary>
    public enum ProcessStatus
    {
        /// <summary>
        /// Plugin initialization in progress
        /// </summary>
        SFILT_STARTING = 0,

        /// <summary>
        /// Plugin is attempting to do its job
        /// </summary>
        SFILT_RUNNING = 1,

        /// <summary>
        /// Plugin successfully completed its job
        /// </summary>
        SFILT_COMPLETE = 2,

        /// <summary>
        /// There was an error somewhere
        /// </summary>
        SFILT_ERROR = 3,

        /// <summary>
        /// An ABORT command has been received; plugin shutdown in progress
        /// </summary>
        SFILT_ABORTING = 4
    }
    #endregion

    #region "Structures"
    
    /// <summary>
    /// Initialization parameters for classes that implement ISpectraFilter
    /// </summary>
    public struct SpectraFilterParams
    {
        public string SourceFolderPath;
        public string OutputFolderPath;
        //' Unused: public Specialized.StringDictionary MiscParams;
        public int DebugLevel;
        public PRISM.Logging.ILogger Logger;
        public IMgrParams MgrParams;
        public IJobParams JobParams;
        public IStatusFile StatusTools;
    }
    #endregion

    /// <summary>
    /// Defines minimum required functionality for classes that will filter spectra files
    /// </summary>
    public interface ISpectraFilter
    {

        #region "Properties"

        /// <summary>
        /// Allows calling program to get current status
        /// </summary>
        ProcessStatus Status { get; }

        /// <summary>
        /// Allows calling program to determine if DTA creation succeeded
        /// </summary>
        ProcessResults Results { get; }

        /// <summary>
        /// Error message describing any errors encountered
        /// </summary>
        string ErrMsg { get; }

        /// <summary>
        /// Count of spectra files that remain after filtering
        /// </summary>
        int SpectraFileCount { get; }
        #endregion

        #region "Methods"

        /// <summary>
        /// Initializes parameters. Must be called before executing Start()
        /// </summary>
        /// <param name="InitParams"></param>
        void Setup(SpectraFilterParams InitParams);

        /// <summary>
        /// Starts the spectra filter operation
        /// </summary>
        /// <returns></returns>
        ProcessStatus Start();

        /// <summary>
        /// //
        /// </summary>
        /// <returns></returns>
        ProcessStatus Abort();
        
        #endregion

    }
}